snuba/utils/streams/profiler.py

```diff
@@ -0,0 +1,70 @@
+import logging
+import time
+from cProfile import Profile
+from pathlib import Path
+from typing import Callable, Mapping, Optional
+
+from snuba.utils.streams.processing import ProcessingStrategy, ProcessingStrategyFactory
+from snuba.utils.streams.types import Message, Partition, TPayload
+
+
+logger = logging.getLogger(__name__)
+
+
+class ProcessingStrategyProfilerWrapper(ProcessingStrategy[TPayload]):
+    def __init__(
+        self,
+        strategy: ProcessingStrategy[TPayload],
+        profiler: Profile,
+        output_path: str,
+    ) -> None:
+        self.__strategy = strategy
+        self.__profiler = profiler
+        self.__output_path = output_path
+
+    def poll(self) -> None:
+        self.__strategy.poll()
+
+    def submit(self, message: Message[TPayload]) -> None:
+        self.__strategy.submit(message)
+
+    def close(self) -> None:
+        self.__strategy.close()
+
+    def terminate(self) -> None:
+        self.__strategy.terminate()
+        self.__profiler.disable()  # not sure if necessary
+        logger.info(
+            "Writing profile data from %r to %r...", self.__profiler, self.__output_path
+        )
+        self.__profiler.dump_stats(self.__output_path)
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__strategy.join(timeout)
+        self.__profiler.disable()  # not sure if necessary
+        logger.info(
+            "Writing profile data from %r to %r...", self.__profiler, self.__output_path
+        )
+        self.__profiler.dump_stats(self.__output_path)
```

Comment:
Do we care if we get an access error at this point because the directory no longer exists or isn't writable?

Reply:
It's obviously not ideal but I don't think there's much we can do if the directory goes away or its permissions are changed between strategy instantiation time and shutdown. I can't think of any major issues with raising here (the strategy is already shut down anyway so it shouldn't be leaving any resources hanging) and I don't think there's any reason to suppress the exception out of caution.

Checking the directory mode at startup might be a good idea to make a best-effort attempt to ensure the directory is writable.